TestModels/Constraints/test/WrappedSimpleConstraintsTest.dfy

```diff
@@ -4,11 +4,565 @@ include "../src/WrappedSimpleConstraintsImpl.dfy"
-    }
+  import opened WrappedSimpleConstraintsService
+  import opened SimpleConstraintsTypes
+  import opened Helpers
+  import SimpleConstraintsImplTest
+  import opened Wrappers
+  import opened StandardLibrary.UInt
+  method{:test} TestConstraints() {
+    var client :- expect WrappedSimpleConstraintsService.WrappedSimpleConstraints();
+    TestGetConstraintWithValidInputs(client);
+    TestGetConstraintWithMyString(client);
+    TestGetConstraintWithOneToTen(client);
+    TestGetConstraintWithTenToTen(client);
+    TestGetConstraintWithLessThanTen(client);
+    TestGetConstraintWithNonEmptyString(client);
+    TestGetConstraintWithStringLessThanOrEqualToTen(client);
+    TestGetConstraintWithMyBlob(client);
+    TestGetConstraintWithNonEmptyBlob(client);
+    TestGetConstraintWithBlobLessThanOrEqualToTen(client);
+    TestGetConstraintWithMyList(client);
+    TestGetConstraintWithNonEmptyList(client);
+    TestGetConstraintWithListLessThanOrEqualToTen(client);
+    TestGetConstraintWithMyMap(client);
+    TestGetConstraintWithNonEmptyMap(client);
+    TestGetConstraintWithMapLessThanOrEqualToTen(client);
+    TestGetConstraintWithGreaterThanOne(client);
+    TestGetConstraintWithUtf8Bytes(client);
+    TestGetConstraintWithListOfUtf8Bytes(client);
+  }
+
+  method TestGetConstraintWithValidInputs(client: ISimpleConstraintsClient)
+    requires client.ValidState()
+    modifies client.Modifies
+    ensures client.ValidState()
+  {
+    var input := GetValidInput();
+    var ret := client.GetConstraints(input := input);
+    expect ret.Success?;
+  }
+
+  // @length(min: 1, max: 10)
+  method TestGetConstraintWithMyString(client: ISimpleConstraintsClient)
+    requires client.ValidState()
+    modifies client.Modifies
+    ensures client.ValidState()
+  {
+    var input := GetValidInput();
+    input := input.(MyString := Some(ForceMyString("this string is really way too long")));
+    var ret := client.GetConstraints(input := input);
+    expect ret.Failure?;
+
+    input := input.(MyString := Some(ForceMyString("12345678901")));
+    ret := client.GetConstraints(input := input);
+    expect ret.Failure?;
+
+    input := input.(MyString := Some(ForceMyString("1234567890")));
+    ret := client.GetConstraints(input := input);
+    expect ret.Success?;
+
+    input := input.(MyString := Some(ForceMyString("1")));
+    ret := client.GetConstraints(input := input);
+    expect ret.Success?;
+
+    input := input.(MyString := Some(ForceMyString("")));
+    ret := client.GetConstraints(input := input);
+    expect ret.Failure?;
+  }
+
+  // @range(min: 1, max: 10)
+  method TestGetConstraintWithOneToTen(client: ISimpleConstraintsClient)
+    requires client.ValidState()
+    modifies client.Modifies
+    ensures client.ValidState()
+  {
+    var input := GetValidInput();
+    input := input.(OneToTen := Some(ForceOneToTen(1000)));
+    var ret := client.GetConstraints(input := input);
+    expect ret.Failure?;
```

Comment:
Nit/Issue: I keep running into tests where we just `expect ret.Failure?` and do not validate what caused the failure or what the failure was.

These tests sometimes mask bugs.
As an example, there is a test that asserts that if no Branch Key ID exists,
the Keystore's Get* methods for that Branch Key ID will fail.

And the Keystore does fail! 
But with a bizarre KMS Exception that no customer could use to realize they were asking for a non-existance Branch Key ID.

Is it possible to bulk replace `ret.Failure?` here with something like `ret.Failure?` AND `ret.Failure == IllegalInputException`?

Reply:
I agree completely, but unfortunately, the error is an Opaque error that came from an exception, and there's no way (as far as I know) to get the actual exception type or the error text out of an Opaque error.
Further, things will likely be different between Java and Dotnet, even if we could get at that stuff; since the errors are generated from custom native code.

Reply:
I just logged in to write the same thing. 
Which is a pretty severe limitation of this Test Model approach we have endorsed. 

@seebees And @robin-aws  for awareness. 

Not that we are going to take any action to resolve this, but we should bear in mind that our limited error expressions mixed with our wrapping test strategy ensures we can never assert the exception type.